dot/parachain/bitfield-signing/bitfield_signing.go

```diff
@@ -0,0 +1,273 @@
+	time.Sleep(availabilityDistributionWaitingPeriod)
+
+	// construct the bitfield according to the availability store
+	bitfield, err := constructAvailabilityBitfield(ctx, rt, validatorIndex, b.subSystemToOverseer)
+	if err != nil {
+		return fmt.Errorf("construct availabilityBitfield: %w", err)
+	}
+
+	// signing process
+	data, err := bitfield.MarshalSCALE()
+	if err != nil {
+		return fmt.Errorf("marshal bitfield for signing: %w", err)
+	}
+	signature, err := validator.Sign(b.keystore, data)
+	if err != nil {
+		return fmt.Errorf("signing bitfield: %w", err)
+	}
+
+	// distribute to subsystem to overseer chan
+	b.subSystemToOverseer <- parachaintypes.DistributeBitfield{
+		RelayParent: relayParent,
+		Bitfield: parachaintypes.UncheckedSignedAvailabilityBitfield{
+			Payload:        bitfield,
+			ValidatorIndex: validatorIndex,
+			Signature:      *signature,
+		},
+	}
+
+	return nil
+}
+
+// ProcessBlockFinalizedSignal processes block finalized signal
+func (b *BitfieldSigning) ProcessBlockFinalizedSignal(parachaintypes.BlockFinalizedSignal) error {
+	return nil
+}
+
+// Stop stops the BitfieldSigning subsystem
+func (b *BitfieldSigning) Stop() {
+	// TODO: anything to do to stop this subsystem ?
+	logger.Infof("Stopping BitfieldSigning subsystem")
+}
+
+func constructAvailabilityBitfield(
+	ctx context.Context,
+	rt runtime.Instance,
+	validatorIdx parachaintypes.ValidatorIndex,
+	subSystemToOverseer chan<- any,
+) (parachaintypes.BitVec, error) {
+	cores, err := rt.ParachainHostAvailabilityCores()
+	if err != nil {
+		return parachaintypes.BitVec{}, fmt.Errorf("querying availability cores: %w", err)
+	}
+
+	// init a bitfield without caring the order
+	bitfield := make([]bitfieldData, 0, len(cores))
+
+	var wg sync.WaitGroup
+	for coreOrderIndex, core := range cores {
+		coreValueIndex, value, err := core.IndexValue()
+		if err != nil {
+			return parachaintypes.BitVec{}, err
+		}
+		// 0 is type of parachaintypes.OccupiedCore
+		if coreValueIndex == 0 {
+			wg.Add(1)
+			go func(oi int, v parachaintypes.OccupiedCore) {
+				defer wg.Done()
+
+				receivingChan := make(chan bool)
+				queryPayload := availabilitystore.QueryChunkAvailability{
+					CandidateHash:  parachaintypes.CandidateHash{Value: v.CandidateHash},
+					ValidatorIndex: uint32(validatorIdx),
+					Sender:         receivingChan,
+				}
+
+				// send QueryChunkAvailability to availability store via overseer
+				subSystemToOverseer <- queryPayload
+
+				// check if cancel signal is coming
+				// we can't cancel the process anytime during the spawned goroutines
+				// instead we only check cancel signal right before receiving from overseer in case it's blocked
+				select {
+				case <-ctx.Done():
+					logger.Infof("process for handleActiveLeavesUpdate is abort due to deactivated leaves")
+					return
+				default:
+					// append the result to the bitfield
+					bitfield = append(bitfield, bitfieldData{index: oi, data: <-receivingChan})
+				}
```

Comment:
Here, this select will not block, it will check if the context is canceled, if so it returns, however if the context is not canceled it will pass to the default case where listen to the channel and if the context is canceled in this mean time the goroutine will never know because the check is in the "past" so the goroutine can be trapped here waiting even if the context is already done. To fix that we need to have a case for each channel, so the select will block until one of those triggers first.
```suggestion
 				var data bool
				select {
				case <-ctx.Done():
					logger.Infof("process for handleActiveLeavesUpdate is abort due to deactivated leaves")
					return
				case res := <-receivingChan:
				        data = res
				}
				
				// append the result to the bitfield
				bitfield = append(bitfield, bitfieldData{index: oi, data:  data})
```

Reply:
that makes total sense! so we need to hear from either `cancel()` or `receivingChan` at this moment, will change it,  this is very productive comment, thanks @EclesioMeloJunior